File: P/24_Swap_Nodes_in_Pairs.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
//Time: O(N) Space: O(1)
class Solution {
    public ListNode swapPairs(ListNode head) {
        if(head == null || head.next == null)
            return head;
        
        ListNode dummy = new ListNode(-1, head);
        ListNode cur = dummy;
        while(cur.next != null && cur.next.next != null) {
            ListNode swap = cur.next.next;
            cur.next.next = swap.next;
            swap.next = cur.next;
            cur.next = swap;
            cur = cur.next.next;
        }
        return dummy.next;
    }
}

File: Paul/LinkedList/24_swapNodesInPairs.js
// https://leetcode.com/problems/swap-nodes-in-pairs/

/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
const swapPairs = function (head) {
  let dummy = new ListNode();
  let start = dummy;
  let count = 0;
  let temp = null;

  if (!head || !head.next) {
    return head;
  }

  // [1,2,3,4]
  while (head.next) {
    if (count < 1) {
      temp = head.next; // 2
      head.next = temp.next; // 3
      temp.next = head; // 1
      start.next = temp;
      head = temp.next;
      count++;
    } else {
      start.next = head;
      head = head.next;
      count = 0;
    }

    start = start.next;
  }

  return dummy.next;
};

const swapPairs = function (head) {
  let dummy = new ListNode(-1, null);
  let start = dummy;

  if (!head || !head.next) {
    return head;
  }

  // [1,2,3,4]
  while (head && head.next) {
    // swap
    let temp = head.next; // 2
    head.next = temp.next; // 3
    temp.next = head; // 1
    start.next = temp;

    // jump
    start = head; // -1 -> 1
    head = head.next; // 1 -> 3
  }

  return dummy.next;
};

// recursive
const swapPairs = function (head) {
  if (!head || !head.next) {
    return head;
  }

  let first = head;
  let second = head.next;

  // swap
  first.next = swapPairs(second.next);
  second.next = first;

  return second;
};
